src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorProjectItem.cs

```diff
@@ -130,9 +130,11 @@ public string FilePathWithoutExtension
             }
         }
 
+        internal RazorSourceDocument SourceDocument { get; set; }
```

Comment:
Hmm, this is concerning. The intent of RazorProjectItem is to be decoupled from a SourceDocument entirely. Why do you need this here? Typically we just re-create a project item.

Reply:
One of the things that helped with perf was having a `RazorSourceLineCollection` that was backed by the Roslyn `SourceText` document (because that's what we're provided in a source generator). Here's what that looks like: https://github.com/dotnet/aspnetcore/pull/28807/files#diff-04e8fa11176d9e8cdfa5d0edeeb3ba7c1c6d20683201775277a3215a1fcb1a3f. 

If I'm not mistaken, there isn't an API that lets you affect how a RazorProjectItem gets turned in to a RazorSourceDocument. We could create one where the default implementation reads from the stream and produces a source document and the source generator configures it with a custom provider. This one felt easier to implement as we did not have to introduce yet another concept, but you have more context here, so open to suggestions.